File: frontend/src/services/tutorialService.js
// Tutorial service to manage tutorial data and progress
const tutorialData = {
  'sorting': {
    id: 'sorting',
    title: 'Sorting Algorithms Mastery',
    difficulty: 'Beginner',
    prerequisites: ['basic-arrays'],
    estimatedTime: '30 minutes',
    steps: [
      {
        id: 'intro',
        title: 'Introduction to Sorting',
        type: 'explanation',
        content: 'Sorting algorithms are fundamental building blocks in computer science...',
        interactive: true
      },
      {
        id: 'bubble-sort',
        title: 'Bubble Sort',
        type: 'explanation',
        content: 'Bubble sort is a simple sorting algorithm...',
        codeExample: `function bubbleSort(arr) {
  const n = arr.length;
  for (let i = 0; i < n - 1; i++) {
    for (let j = 0; j < n - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        // Swap elements
        [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
      }
    }
  }
  return arr;
}`,
        interactive: true,
        quiz: {
          question: 'What is the time complexity of bubble sort?',
          options: ['O(n)', 'O(n²)', 'O(n log n)', 'O(1)'],
          correct: 1
        }
      }
    ]
  },
  'graph': {
    id: 'graph',
    title: 'Graph Algorithms',
    difficulty: 'Intermediate',
    prerequisites: ['basic-arrays', 'sorting'],
    estimatedTime: '45 minutes',
    steps: [
      {
        id: 'intro',
        title: 'Introduction to Graphs',
        type: 'explanation',
        content: 'Graphs are versatile data structures...',
        interactive: true
      }
    ]
  }
};

export const getTutorial = async (id) => {
  // Simulate API call
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(tutorialData[id]);
    }, 500);
  });
};

export const getTutorialProgress = (tutorialId) => {
  const progress = localStorage.getItem('tutorialProgress');
  if (progress) {
    const allProgress = JSON.parse(progress);
    return allProgress[tutorialId];
  }
  return null;
};

export const saveTutorialProgress = (tutorialId, progress) => {
  const existingProgress = localStorage.getItem('tutorialProgress');
  const allProgress = existingProgress ? JSON.parse(existingProgress) : {};
  
  allProgress[tutorialId] = progress;
  localStorage.setItem('tutorialProgress', JSON.stringify(allProgress));
};

export const generateCertificate = (tutorialId) => {
  const progress = getTutorialProgress(tutorialId);
  if (!progress || !tutorialData[tutorialId]) {
    return null;
  }

  const tutorial = tutorialData[tutorialId];
  const completionPercentage = (progress.completedSteps.length / tutorial.steps.length) * 100;

  if (completionPercentage < 100) {
    return null;
  }

  return {
    tutorialTitle: tutorial.title,
    completionDate: new Date().toISOString(),
    difficulty: tutorial.difficulty,
    studentName: 'Student Name' // TODO: Add user management
  };
};

export default {
  getTutorial,
  getTutorialProgress,
  saveTutorialProgress,
  generateCertificate
};

File: backend/services/tutorial_service.py
from typing import Dict, List, Optional
from models.tutorial_models import Tutorial, TutorialStep, Quiz, QuizOption

# Mock database of tutorials
TUTORIALS_DB: Dict[str, Tutorial] = {
    "sorting": Tutorial(
        id="sorting",
        title="Sorting Algorithms Mastery",
        description="Learn about different sorting algorithms and their implementations",
        difficulty="Beginner",
        estimatedTime="30 minutes",
        prerequisites=["basic-arrays"],
        category="sorting",
        icon="🔄",
        steps=[
            TutorialStep(
                id="intro",
                type="explanation",
                title="Introduction to Sorting",
                content="Sorting is a fundamental operation in computer science...",
                interactive=True
            ),
            TutorialStep(
                id="bubble-sort",
                type="explanation",
                title="Bubble Sort Algorithm",
                content="Bubble Sort is the simplest sorting algorithm...",
                codeExample="""def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr""",
                interactive=True,
                quiz=Quiz(
                    question="What is the time complexity of Bubble Sort?",
                    options=[
                        QuizOption(text="O(n)", isCorrect=False),
                        QuizOption(text="O(n²)", isCorrect=True),
                        QuizOption(text="O(n log n)", isCorrect=False),
                        QuizOption(text="O(1)", isCorrect=False)
                    ],
                    explanation="Bubble Sort uses nested loops, leading to O(n²) time complexity"
                )
            )
        ]
    ),
    "graph": Tutorial(
        id="graph",
        title="Graph Algorithms",
        description="Master graph traversal and pathfinding algorithms",
        difficulty="Intermediate",
        estimatedTime="45 minutes",
        prerequisites=["basic-arrays", "sorting"],
        category="graph",
        icon="🕸️",
        steps=[
            TutorialStep(
                id="intro",
                type="explanation",
                title="Introduction to Graphs",
                content="Graphs are versatile data structures...",
                interactive=True
            )
        ]
    )
}

class TutorialService:
    @staticmethod
    def get_all_tutorials() -> List[Tutorial]:
        """Get all available tutorials."""
        return list(TUTORIALS_DB.values())

    @staticmethod
    def get_tutorial_by_id(tutorial_id: str) -> Optional[Tutorial]:
        """Get a specific tutorial by ID."""
        return TUTORIALS_DB.get(tutorial_id)

    @staticmethod
    def get_tutorials_by_difficulty(difficulty: str) -> List[Tutorial]:
        """Get tutorials filtered by difficulty level."""
        return [
            tutorial for tutorial in TUTORIALS_DB.values()
            if tutorial.difficulty.lower() == difficulty.lower()
        ]

    @staticmethod
    def get_tutorials_by_category(category: str) -> List[Tutorial]:
        """Get tutorials filtered by category."""
        return [
            tutorial for tutorial in TUTORIALS_DB.values()
            if tutorial.category.lower() == category.lower()
        ]
